[JENKINS-59923] Replaced TagAction with new ManifestAction

willi1s opened the pull request:
JENKINS-59923: Repo Manifest action clashes with the built in Tags action

rsandell commented on src/main/java/hudson/plugins/repo/ManifestAction.java:
```diff
@@ -0,0 +1,107 @@
+	 * Returns the name of the Url to use for the action.
+	 */
+	public final String getUrlName() {
+		return "manifest";
```
To be even more secured to future clashes maybe `repo-manifest` would be even better :)

rsandell commented on src/main/java/hudson/plugins/repo/TagAction.java:
```diff
@@ -23,68 +23,50 @@
+    }
 
-	@Override
-	public String getTooltip() {
```
I don't like that you changed the binary signature, we strive for backwards compatibility, and since this class is not going to be used going forward I see no need to remove methods from it. It's fine to return `null` from `getDisplayName` et.al. though :)

willi1s replied:
Understood. I've now reverted most of the changes to this file to maintain the binary signature

rsandell commented on src/main/java/hudson/plugins/repo/TagAction.java:
```diff
@@ -23,68 +23,50 @@
+ * TagAction class. This class contains code to migrate from the old TagAction
+ * to a ManifestAction, as the TagAction (1.10.7) conflicts with the builtin tag code.
  */
 @ExportedBean(defaultVisibility = 999)
```
```suggestion
@Depreacated
```

willi1s replied:
I have updated this as requested

rsandell commented on src/main/java/hudson/plugins/repo/TagAction.java:
```diff
@@ -23,68 +23,50 @@
- * to recreate the exact state of the repository when the build was ran.
+ * TagAction class. This class contains code to migrate from the old TagAction
+ * to a ManifestAction, as the TagAction (1.10.7) conflicts with the builtin tag code.
  */
```
I suggest to keep the original documentation text and just add

```
 * @deprecated replaced by {@link ManifestAction} JENKINS-59923
```

or something...

willi1s replied:
Updated, as requested

rsandell commented on src/main/java/hudson/plugins/repo/TagAction.java:
```diff
@@ -23,68 +23,50 @@
  */
 @ExportedBean(defaultVisibility = 999)
-public class TagAction extends AbstractScmTagAction {
+public class TagAction implements RunAction2 {
```
Try to keep the original binary signature, and change the restriction so noone can use it when they upgrade their dependency.
```suggestion
@Restricted(NoExternalUse.class)
public class TagAction implements AbstractScmTagAction {
```

willi1s replied:
Sounds good to me. I've done this now.

rsandell commented on src/main/java/hudson/plugins/repo/TagAction.java:
```diff
@@ -23,68 +23,50 @@
-	}
+        /* Use the replaceAction method, so if this method gets called more than once,
+           we don't end up with multiple ManifestActions attached */
+        r.replaceAction(new ManifestAction(r));
```
This won't work since `Actionable.addOrReplaceAction` uses `equals` to determine if it is the same action, and the default `equals` implementation just looks at the reference pointer you will get multiple `ManifestActions` anyway.

I suggest to implement a `readResolve` instead that just returns a new `ManifestAction` instance. `load` should be called on that new resolved object IIRC.

willi1s replied:
`readResolve` is new to me. I've updated the code, and retested. It's all working fine, and I think this is a much better solution.

willi1s commented:
> Did you miss to move the old `badge.jelly`?

This file is not required as I'm not extending `AbstractScmTagAction` in the `ManifestAction` class.

willi1s commented:
I have made the requested review updates, and also rebased from the upstream master.

rsandell commented:
Please in the future don't force push, it makes reviewing what has changed since I last reviewed a pain and take me longer to do since I need to re review from scratch.

willi1s commented:
> Please in the future don't force push, it makes reviewing what has changed since I last reviewed a pain and take me longer to do since I need to re review from scratch.

Apologies. I'm new to github, so not too sure of the correct process. I'll be sure not to force in future.

rsandell commented:
> This file is not required as I'm not extending AbstractScmTagAction in the ManifestAction class.

But that means that you have removed a feature (the badge in the build history widget) you can add it back and implement `hudson.model.BuildBadgeAction` to get the icon back.

willi1s commented:
@rsandell Sorry for the misunderstanding, I hadn't spotted that feature.
I have added an updated badge.jelly and implemented the `BuildBadgeAction` and have tested it as working on my local Jenkins.

willi1s commented:
> The merge looks weird, maybe because you force pushed in the past I don't know.
> I'll do some local merging first to see nothing gets screwed up.

Sorry if I made it a bit more painful than it could have been. 
Thanks for your support.
